src/main/kotlin/com/demonwav/mcdev/platform/mcp/debug/UngrabMouseDebugSessionListener.kt

```diff
@@ -30,18 +31,58 @@ class UngrabMouseDebugSessionListener(private val process: DebugProcessImpl) : X
         val virtualMachine = debugProcess.virtualMachineProxy as? VirtualMachineProxyImpl ?: return
         val evaluationContext = EvaluationContextImpl(suspendContextImpl, frameProxy)
 
-        val mouseClass = virtualMachine.classesByName("org.lwjgl.input.Mouse")?.singleOrNull() as? ClassType ?: return
+        val mouseClass = virtualMachine.classesByName("org.lwjgl.input.Mouse")?.singleOrNull() as? ClassType
+        // LWJGL 3 does not have the Mouse class, Minecraft uses its own MouseHelper instead
+        if (mouseClass != null) {
+            ungrab2(mouseClass, virtualMachine, debugProcess, evaluationContext)
+        } else {
+            ungrab3(virtualMachine, debugProcess, evaluationContext)
+        }
+    }
+
+    private fun ungrab2(
+        mouseClass: ClassType,
+        virtualMachine: VirtualMachineProxyImpl,
+        debugProcess: DebugProcessImpl,
+        evaluationContext: EvaluationContextImpl
+    ) {
         val isGrabbed = mouseClass.methodsByName("isGrabbed", "()Z")?.singleOrNull() ?: return
         val setGrabbed = mouseClass.methodsByName("setGrabbed", "(Z)V")?.singleOrNull() ?: return
-        val grabValue = virtualMachine.mirrorOf(grab)
+        val grabValue = virtualMachine.mirrorOf(false)
 
         val currentState = debugProcess.invokeMethod(evaluationContext, mouseClass, isGrabbed, emptyList())
-        if ((currentState as? BooleanValue)?.value() != grab) {
+        if ((currentState as? BooleanValue)?.value() == true) {
             debugProcess.invokeMethod(evaluationContext, mouseClass, setGrabbed, listOf(grabValue))
         }
     }
 
+    private fun ungrab3(
+        virtualMachine: VirtualMachineProxyImpl,
+        debugProcess: DebugProcessImpl,
+        evaluationContext: EvaluationContextImpl
+    ) {
+        val minecraftClass =
+            virtualMachine.classesByName("net.minecraft.client.Minecraft")?.singleOrNull() as? ClassType ?: return
+        val minecraftGetter =
+            minecraftClass.methodsByName("getInstance", "()Lnet/minecraft/client/Minecraft;")?.singleOrNull() ?: return
+        val minecraft = debugProcess.invokeMethod(
+            evaluationContext,
+            minecraftClass,
+            minecraftGetter,
+            emptyList()
+        ) as? ObjectReference ?: return
+
+        val mouseHelperField = minecraftClass.fieldByName("mouseHelper") ?: return
+        val mouseHelper = minecraft.getValue(mouseHelperField) as? ObjectReference ?: return
+
+        val mouseHelperClass =
+            virtualMachine.classesByName("net.minecraft.client.MouseHelper")?.singleOrNull() as? ClassType ?: return
+        val grabMouse = mouseHelperClass.methodsByName("ungrabMouse", "()V")?.singleOrNull() ?: return
```

Comment:
Maybe it would be better to not hardcode mcp method names, and try to map srg names based on current mappings if possible? (and fall back to hardcoded name if mappings not found)

Reply:
I thought about it, I don't know how to get the relevant module though.

Reply:
There are 2 ways to do it that I can see, the easy way, and the way I think you should do it:

_The easy way_: You can get the current `Project` from the `process` field in `UngrabMouseDebugSessionListener` and use that to find _some_ `SrgManager` in the project:

```
SrgManager.findAnyInstance(process.project)
```

 I added that method specifically for cases like this where we don't have a `Module` to go off of....except

----

_The way I think you should do it_: We do know the module we're working with, we just need to move the code around a bit.

`McpRunConfigurationExtension` is what creates the `UngrabMouseDebugSessionListener` instance and registers it as a listener, and it _does_ know which `Module` it's working with. That class extends both `ModuleDebugRunConfigurationExtension` and `DebugProcessListener`, the first does know about `Module`s, the second does not.

So to make this work those 2 classes will need to be split up as separate implementations and [here](https://github.com/minecraft-dev/MinecraftDev/blob/9fbc8c8d9d4cec0126ede2d2f2b37ed4b6653907/src/main/kotlin/com/demonwav/mcdev/platform/mcp/debug/McpRunConfigurationExtension.kt#L27) instead of passing in `this` it will need to pass in a new object that implements `DebugProcessListener` and contains a field referencing that `Module` so it can pass that `Module` to `UngrabMouseDebugSessionListener` in `processAttached`.

With that done you can get the `SrgManager` with:

```
MinecraftFacet.getInstance(module, McpModuleType)?.srgManager
```

But also keep the MCP names in as a fallback in case a SRG mapping can't be found (happens frequently enough considering how finicky FG is).